### PR TITLE
feat: people merge improvements

### LINKED ISSUE #31392: Merging a named person into a person without a name silently fails
### I have searched the existing issues, both open and closed, to make sure this is not a duplicate report.

- [x] Yes

### Description
When naming a person, if it finds a duplicate name, you are given the option to merge the two people together. If you merge the unnamed person into the named person it works fine. However, if you click the switch direction icon, this then tries to merge the named person into the unnamed person. The request comes back as a 200, and you will see a success dialog. However, it fails on the backend early due to a missing name.

### The OS that Immich Server is running on

Windows 11 (Docker Desktop, WSL2 backend)

### Version of Immich Server

v3-rc (but verified the code is the same on main)

### Version of Immich Mobile App

n/a

### Platform with the issue

- [x] Server
- [ ] Web
- [ ] Mobile

### Device make and model

Macbook chrome browser

### Your docker-compose.yml content

```YAML
name: immich
services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: 1
              capabilities: [gpu]
    volumes:
      - G:/ImmichUpload/upload:/data/upload
      - G:/ImmichUpload/library:/data/library
      - G:/ImmichUpload/backups:/data/backups
      - D:/ImmichData/thumbs:/data/thumbs
      - D:/ImmichData/encoded-video:/data/encoded-video
      - D:/ImmichData/profile:/data/profile
      - /etc/localtime:/etc/localtime:ro
      - G:/:/mnt/family-photos
    env_file:
      - .env
    ports:
      - '2283:2283'
    depends_on:
      - redis
      - database
    restart: always
    healthcheck:
      disable: false
  immich-machine-learning:
    container_name: immich_machine_learning
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}-cuda
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              count: 1
              capabilities: [gpu]
    volumes:
      - model-cache:/cache
    env_file:
      - .env
    restart: always
    healthcheck:
      disable: false
  redis:
    container_name: immich_redis
    image: docker.io/valkey/valkey:9@sha256:8e8d64b405ce18f41b8e5ee20aa4687a8ed0022d1298f2ce31cdcf3a76e09411
    healthcheck:
      test: redis-cli ping || exit 1
    restart: always
  database:
    container_name: immich_postgres
    image: ghcr.io/immich-app/postgres:14-vectorchord0.4.3-pgvectors0.2.0@sha256:bcf63357191b76a916ae5eb93464d65c07511da41e3bf7a8416db519b40b1c23
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
      POSTGRES_INITDB_ARGS: '--data-checksums'
    volumes:
      - ${DB_DATA_LOCATION}:/var/lib/postgresql/data
    shm_size: 128mb
    restart: always
    healthcheck:
      disable: false
  cloudflared:
    container_name: immich_cloudflared
    image: cloudflare/cloudflared:latest
    command: tunnel --no-autoupdate run --token ${CLOUDFLARE_TUNNEL_TOKEN}
    restart: always
volumes:
  model-cache:
```

### Your .env content

```Shell
DB_DATA_LOCATION=C:/immich-app/postgres
TZ=America/Denver
IMMICH_VERSION=v3-rc
DB_PASSWORD=REDACTED
CLOUDFLARE_TUNNEL_TOKEN=REDACTED
DB_USERNAME=postgres
DB_DATABASE_NAME=immich
```

### Reproduction steps

1. Have two people entries that are really the same person. Person A has a name, like "John". Person B doesn't have a name yet.
2. Open up Person B (the one without a name)
3. Click to add their name
4. Type in the same name as Person A, "John".
5. A dialog asks if this is the same person, say yes.
6. Click the double arrow icon to switch the direction of the merge
7. Click merge
8. Success popup shows that the merge and the rename happaned
9. Notice that you still have a person without a name

You can also reproduce this through the API:
`POST /api/people/{personB_id}/merge
{ "ids": ["{personA_id}"] }`

Response body: []

Expected Behavior
The unnamed person should have a name (I didn't test this with birthdate, but I'd expect that to copy over as well). The merge should succeed.

### Relevant log output

```shell

```

### Root cause
/server/src/services/person.service.ts
```
      const changes: Updateable<Person> = {};
      if (!primaryPerson.name && mergePerson.name) {
        changes.name = mergePerson.name;
      }

      if (!primaryPerson.birthDate && mergePerson.birthDate) {
        changes.birthDate = mergePerson.birthDate;
      }

      if (
        (mergePerson.name && mergePerson.name !== primaryPerson.name) ||
        (mergePerson.birthDate && mergePerson.birthDate !== primaryPerson.birthDate)
      ) {
        continue; // this breaks us out of the loop without changing anything
      }
```

Note that if the name doesn't match, this would also fail. Like if I had a nickname in one and tried to merge into one with their real name. But I'd be okay if this just worked for blank names.

### PR DESCRIPTION
**Pull request by danieldietzler**
Fixes #31392

### CODE REVIEW

File: server/src/services/person.service.ts [hunk 1]
```diff
@@ -578,82 +578,82 @@ export class PersonService extends BaseService {
...
       if (!hasAccess) {
         results.push({ id: mergeId, success: false, error: BulkIdErrorReason.NO_PERMISSION });
         continue;
       }
 
-      if (!primaryPerson || primaryPerson.ownerId !== mergePerson.ownerId) {
-        primaryPerson = await this.personRepository.getByGroupId({ ownerId: mergePerson.ownerId, personGroupId });
-        if (!primaryPerson) {
-          continue;
+      if (!mergedPerson || mergedPerson.ownerId !== mergePerson.ownerId) {
```
Comment (jrasm91) [inline server/src/services/person.service.ts]: do you not have the right information to just use `mergePerson` directly instead of re-querying it again?

File: server/src/services/person.service.ts [hunk 2]
```diff
@@ -578,82 +578,82 @@ export class PersonService extends BaseService {
...
+          ownerId: mergePerson.ownerId,
+          personGroupId: mergePerson.personGroupId,
+        });
+        continue;
+      }
+
+      if (mergedPerson.name !== mergePerson.name || mergedPerson.birthDate !== mergePerson.birthDate) {
+        if (mergePerson.ownerId === auth.user.id) {
+          throw new BadRequestException('Cannot merge two people with different names or birth dates');
         }
```
Comment (jrasm91) [inline server/src/services/person.service.ts]: This is not true and we should not error when this happens. We should use the order of the ids to determine the final name and birthdate of the people.

File: server/src/services/person.service.ts [hunk 3]
```diff
@@ -578,82 +578,82 @@ export class PersonService extends BaseService {
...
-      permission: Permission.PersonMerge,
-      ids: mergeIds,
-    });
+    const allowedIds = await this.checkAccess({ auth, permission: Permission.PersonMerge, ids });
 
-    let primaryPerson: Selectable<PersonTable> | undefined;
+    let mergedPerson: Selectable<PersonTable> | undefined;
 
-    for (const mergePerson of await this.personRepository.getForMergePerson(mergeIds)) {
+    for (const mergePerson of await this.personRepository.getForMergePerson(ids)) {
```
Comment (jrasm91) [inline server/src/services/person.service.ts:597]: it should order based off of the dto.ids not random postgres return order

File: server/src/services/person.service.ts [hunk 4]
```diff
@@ -578,82 +578,82 @@ export class PersonService extends BaseService {
...
 
-    const allowedIds = await this.checkAccess({
-      auth,
-      permission: Permission.PersonMerge,
-      ids: mergeIds,
-    });
+    const allowedIds = await this.checkAccess({ auth, permission: Permission.PersonMerge, ids });
 
-    let primaryPerson: Selectable<PersonTable> | undefined;
+    let mergedPerson: Selectable<PersonTable> | undefined;
```
Comment (jrasm91) [inline server/src/services/person.service.ts]: might be better to name this `targetPerson` or `mainPerson` instead. They, in fact, aren't merged into anything they're being merged into and that seems a bit ambiguous.

**Review by jrasm91 — CHANGES_REQUESTED** [review]

File: server/src/controllers/person.controller.ts [hunk 5]
```diff
@@ -187,19 +187,35 @@ export class PersonController {
     return this.service.reassignFaces(auth, id, dto);
   }
 
+  @Post('merge')
+  @Authenticated({ permission: Permission.PersonMerge })
+  @HttpCode(HttpStatus.OK)
+  @Endpoint({
+    summary: 'Merge people',
+    description: 'Merge a list of people into the person specified in the path parameter.',
```
Comment (trcoop) [inline server/src/controllers/person.controller.ts]: This isn't true anymore, they get merged into the first person in the list of people. I'd actually argue that this isn't an improvement. When you had the primary person on the path, that was self documenting who was going to remain, this is now more vague and the description doesn't help clarify it. I'm curious what problem the previous api call had that is trying to be solved here, because this also now means you have to support both until you can fully deprecate the other one.

File: server/src/services/person.service.ts [hunk 6]
```diff
@@ -578,82 +578,82 @@ export class PersonService extends BaseService {
...
-    if (mergeIds.includes(personGroupId)) {
-      throw new BadRequestException('Cannot merge a person into themselves');
+  async mergePeople(auth: AuthDto, { ids }: MergePersonDto): Promise<BulkIdResponseDto[]> {
+    if (ids.length < 2) {
+      throw new BadRequestException('At least two people are required for merging');
     }
 
-    await this.requireAccess({ auth, permission: Permission.PersonUpdate, ids: [personGroupId] });
+    if (new Set(ids).size !== ids.length) {
+      throw new BadRequestException('Cannot merge a person into themselves');
```
Comment (trcoop) [inline server/src/services/person.service.ts:588]: Technically the list of other people to merge could have had a duplicate, so this won't always be true. For exampe, you might have ["Person A", "Person B", "Person B"]. Previously that wouldn't have failed here, we would have merged the first person in, but then when trying to merge the second person that would have failed, so we would have got a partial merge. Now we are just failing outright.

I do think this is probably an extreme edge case though, I don't think the UI would even allow it, but something to consider. Might be worth adding a test case that shows this, so it is known that duplicates later in the list are also going to fail everything.

File: server/src/services/person.service.ts (hunk 4 above)
Comment (trcoop) [inline server/src/services/person.service.ts]: I know this is a nit, and I really hate it when people leave nits, but...

I had to read over the code multiple times to get oriented because it was really hard to tell the difference between mergePerson and mergedPerson. It was a lot easier to read with primaryPerson so the variable names didn't look quite so similar.

Again, sorry for commenting on a nit, but it would help with readability, and thus future mantainability.

File: server/src/services/person.service.ts [hunk 7]
```diff
@@ -578,82 +578,82 @@ export class PersonService extends BaseService {
...
-          continue;
+      if (!mergedPerson || mergedPerson.ownerId !== mergePerson.ownerId) {
+        mergedPerson = await this.personRepository.getByGroupId({
+          ownerId: mergePerson.ownerId,
+          personGroupId: mergePerson.personGroupId,
+        });
+        continue;
+      }
+
+      if (mergedPerson.name !== mergePerson.name || mergedPerson.birthDate !== mergePerson.birthDate) {
```
Comment (trcoop) [inline server/src/services/person.service.ts]: This just went backwards from what I reported rather than fixing it. We used to be able to merge No Name->Name. But not Name->No Name. Now we can't merge regardless of which side the blank name is on.

Comment (jrasm91) [reply server/src/services/person.service.ts]: Nah, this was already updated to always merge all requested ids, regardless of data.

File: server/src/services/person.service.ts [hunk 8]
```diff
@@ -578,82 +578,82 @@ export class PersonService extends BaseService {
...
+        if (mergePerson.ownerId === auth.user.id) {
```
Comment (trcoop) [inline server/src/services/person.service.ts]: I'm still confused on the business logic of people with different owners.

Reading this code I think it is saying that if there are different owners, but they both use the same name, we can merge them. But if they are different names we'll skip if the owners aren't the same, but we'll error if the owners are the same.

Why would we even be trying to merge two names from different owners? I don't even know how to do that. It would have to be somebody making an API call to get into that situation, and I don't know the business case for why we would allow that. I feel like it would be cleaner to error out if all the people to merge aren't owned by the same person. I think I'm just missing some context/history here.

File: server/test/medium/specs/services/person.service.spec.ts [hunk 9]
```diff
@@ -266,7 +266,27 @@ describe(PersonService.name, () => {
       );
     });
 
-    it('should skip people with a different birthdate', async () => {
+    it('should error on people with a different name that both belong to the auth user', async () => {
```
Comment (trcoop) [inline server/test/medium/specs/services/person.service.spec.ts]: What should happen if they have a different name and belong to a different user? I think having that test case would help to start documenting why the group matters.

File: server/src/services/person.service.ts [hunk 10]
```diff
@@ -578,82 +578,90 @@ export class PersonService extends BaseService {
...
+    }
 
-      if (!primaryPerson || primaryPerson.ownerId !== mergePerson.ownerId) {
-        primaryPerson = await this.personRepository.getByGroupId({ ownerId: mergePerson.ownerId, personGroupId });
-        if (!primaryPerson) {
+    // eslint-disable-next-line unicorn/prefer-object-iterable-methods
+    for (const ownerId of Object.keys(ownerPeopleMap)) {
+      let targetPerson = ownerPeopleMap[ownerId][ids[0]];
+
+      for (const mergeId of ids.slice(1)) {
```
Comment (trcoop) [inline server/src/services/person.service.ts]: This slice is on the original list of ids, meaning you'll get the full list (minus the first id) from the request for each owner map you are iterating over.

If you have ids that have different owners, I think you'll get an exception on line 615 when you try to access the ownerId property.

File: server/src/services/person.service.ts [hunk 11]
```diff
@@ -578,82 +578,90 @@ export class PersonService extends BaseService {
...
+        ownerPeopleMap[mergePerson.ownerId] = {};
       }
+      ownerPeopleMap[mergePerson.ownerId][mergePerson.personGroupId] = mergePerson;
+    }
 
-      if (!primaryPerson || primaryPerson.ownerId !== mergePerson.ownerId) {
-        primaryPerson = await this.personRepository.getByGroupId({ ownerId: mergePerson.ownerId, personGroupId });
-        if (!primaryPerson) {
+    // eslint-disable-next-line unicorn/prefer-object-iterable-methods
+    for (const ownerId of Object.keys(ownerPeopleMap)) {
```
Comment (trcoop) [inline server/src/services/person.service.ts]: Okay, I think I'm starting to get the picture now of what this is trying to do. You're saying that you want to support multiple ids coming in from multiple owners, and then only merge the ids with in the same owner.

So, if we have Owner A ids: "p1-A", "p2-A". Then Owner B ids: "p3-B", "p4-B", then you expect p2-A to merge into p1-a and p4-B to merge in to p3-B?

So, I'd pass the ids in the body as: ["p1-A", "p2-A", "p3-B", "p4-B"], then I'd have an end result of p2-a and p4-B surviving?  That's what I think I'm gathering from your intent here.

I didn't realize you would be trying to merge multiple people from multiple owners at the same time. Admitedly, I don't understand the group logic well enough. I thought they shared the same group, so merging one person into another would just automatically fix it for everybody in the group.

If this is the intent, a unit test showing that example would be great to see.

File: server/src/services/person.service.ts [hunk 12]
```diff
@@ -578,82 +578,90 @@ export class PersonService extends BaseService {
...
+      let targetPerson = ownerPeopleMap[ownerId][ids[0]];
+
+      for (const mergeId of ids.slice(1)) {
+        const mergePerson = ownerPeopleMap[ownerId][mergeId];
+        const hasAccess = allowedIds.has(mergeId);
```
Comment (trcoop) [inline server/src/services/person.service.ts]: You check for access on the mergePerson, but never check for access on the targetPerson.

File: server/src/services/person.service.ts (hunk 11 above)
Comment (trcoop) [reply server/src/services/person.service.ts]: Okay, I finally see it. This is just hard to visualize looking at only the PR.

I pass in ["p1-A", "p2-A"]. The fetch sees this is part of a group, and then pulls in "p3-B" and "p4-B". We then separate those out into maps based on their owners.

Then we get "p1-A" and compare it to "p2-A".
"p1-A" has the same owner as the auth id, so we merge no matter what.

Then we get "p3-B" and compare it to "p4-B".
"p3-B" doesn't have the same owner as auth, so we only merge if the names and birth dates match.

Okay, this is making more sense now.

File: server/src/services/person.service.ts [hunk 13]
```diff
@@ -578,82 +578,90 @@ export class PersonService extends BaseService {
...
 
-      if (!primaryPerson || primaryPerson.ownerId !== mergePerson.ownerId) {
-        primaryPerson = await this.personRepository.getByGroupId({ ownerId: mergePerson.ownerId, personGroupId });
-        if (!primaryPerson) {
+    // eslint-disable-next-line unicorn/prefer-object-iterable-methods
+    for (const ownerId of Object.keys(ownerPeopleMap)) {
+      let targetPerson = ownerPeopleMap[ownerId][ids[0]];
+
+      for (const mergeId of ids.slice(1)) {
+        const mergePerson = ownerPeopleMap[ownerId][mergeId];
```
Comment (trcoop) [inline server/src/services/person.service.ts]: mergePerson could be undefined here, if the owner doesn't have a person matching the groupId for this target.

For example, let's say we have 3 people in the same cluster group. Me, Spouse, Child.

For Myself, I have the two people I'm trying to merge personA and personB. My Spouse has personA, but not personB. My Child has both people.

We fetch from the DB, and once the ordering that was mentioned above is addressed, I'd get them all back and stick them into a map like this:

`ownerPeopleMap = {
  me: { P1: <row>, P2: <row> },
  spouse: { P1: <row> },
  child: { P1: <row>, P2: <row> }
}`

We go through my map first and everything is fine.

We go through my spouses map, this mergePerson is undefined, they don't have P2. Line #609 goes through fine, the auth user has access to P2. Then on line #615, we try to dereference on undefined mergePerson and throws an exception. We likely get back an error to the client? I didn't follow all that back up the chain.

Because we failed early, we didn't even process the merge for my child. So we end up with them still having the duplicate.

File: server/src/services/person.service.ts [hunk 14]
```diff
@@ -578,82 +578,89 @@ export class PersonService extends BaseService {
...
-        ownerId: primaryPerson.ownerId,
-      };
-      this.logger.log(`Merging ${mergeName} into ${primaryPerson.name || primaryPerson.personGroupId}`);
+        if (Object.keys(changes).length > 0) {
+          targetPeople[mergePerson.ownerId] = await this.personRepository.update({
+            ownerId: targetPerson.ownerId,
+            personGroupId: targetPerson.personGroupId,
+            ...changes,
+          });
+        }
```
Comment (jrasm91) [inline server/src/services/person.service.ts:642]: this can probably be cleaned up a bit

File: server/src/services/person.service.ts [hunk 15]
```diff
@@ -578,82 +578,89 @@ export class PersonService extends BaseService {
...
 
-      if (!primaryPerson.birthDate && mergePerson.birthDate) {
-        changes.birthDate = mergePerson.birthDate;
-      }
+        if (
+          (targetPerson.name !== mergePerson.name || targetPerson.birthDate !== mergePerson.birthDate) &&
+          mergePerson.ownerId !== auth.user.id
+        ) {
+          continue;
+        }
```
Comment (jrasm91) [inline server/src/services/person.service.ts:625]: make sure the allow merging unnamed into named, even for the non-auth user

File: server/src/controllers/person.controller.ts [hunk 16]
```diff
@@ -187,19 +187,35 @@ export class PersonController {
     return this.service.reassignFaces(auth, id, dto);
   }
 
+  @Post('merge')
+  @Authenticated({ permission: Permission.PersonMerge })
+  @HttpCode(HttpStatus.OK)
+  @Endpoint({
+    summary: 'Merge people',
+    description: 'Merge an ordered list of people.',
+    history: new HistoryBuilder().added('v3.2').stable('v3.2'),
```
Comment (jrasm91) [inline server/src/controllers/person.controller.ts]: added in 3.2.1

**Review by jrasm91 — APPROVED** [review]

**Review by jrasm91 — APPROVED** [review]
